### ISSUE TITLE
[Syntax] Make mutable variables longer and more awkward to specify

### ISSUE THREAD

**Issue by redradist** [modular/max#451]
### Review Mojo's priorities

- [X] I have read the [roadmap and priorities](https://docs.modular.com/mojo/roadmap.html#overall-priorities) and I believe this request falls within the priorities.

### What is your request?
Use for variables only single keyword `let` and not use `var`.
Instead if user need to mutate literal, use the `let mut` syntax like Rust did

### What is your motivation for this change?
Reason is that such feature will enforce users to use more constants by default that will increase performance overall

**Comment by gryznar** [comment]
+1 for this. Current syntax may be also quite confusing if `let` is used only to declare immutable variables and not to declare variables overall

**Comment by drunkwcodes** [comment]
How to eliminate variables?

**Comment by lattner** [comment]
I retitled this, because it doesn't sound like you're suggesting we change how constant values are defined, you're suggesting that we make it harder to write a mutable variable than just "var"?

**Comment by oskgo** [comment]
While this change would make it more awkward to specify mutable variables, it actually makes it less awkward to switch between mutable and immutable.

The downside is that you have to type more characters when you know that it has to be mutable to begin with. The upside is that you're more likely to prefer using the shorter option (immutable) when you don't know.

With the current syntax users are more inclined to default to the more permissible option (mutable) when they don't know.

**Comment by redradist** [comment]
> I retitled this, because it doesn't sound like you're suggesting we change how constant values are defined, you're suggesting that we make it harder to write a mutable variable than just "var"?

@lattner Yes, exactly, just like it was done in Rust

**Comment by redradist** [comment]
@oskgo 
> With the current syntax users are more inclined to default to the more permissible option (mutable) when they don't know.

Yes, and it will make work of optimizer harder. With my proposal users will more likely write shorter name `let` and use variables only then when they really will be needed

**Comment by redradist** [comment]
@drunkwcodes 
> How to eliminate variables?

Variables are not eliminated, they just harder to write with extra `mut`

**Comment by ksandvik** [comment]
To start with using let/var to separate unmutable versus mutable is fine for me. But it makes sense to some degree to make all variables non-mutable and the developer has to think if she/he really needs mutability by adding a keyword or some other language syntax. Plus you get by default the optimizations by immutable vaues. The compiler will flag anyway all cases where an immutable value is mutated.

**Comment by gryznar** [comment]
One more thing to add here is the fact, that compiler currently keeps eye on proper usage of let / var and produces a warning if variable was assigned to `var` if it is immutable. Maybe step forward would be just to replace this warning with error?

Warning could be just ignored, but if someone is using this keywords it will be much better if it would use them correctly

**Comment by drunkwcodes** [comment]
No way.

![20230715_211138](https://github.com/modularml/mojo/assets/36228443/043f2d8d-b252-442c-8e14-14d72c98c69a)

**Comment by gryznar** [comment]
@drunkwcodes you just behave like a usual troll (rejecting ideas without any argument, mocking approach etc.) Keeping you serious is just a misunderstanding.

**Comment by drunkwcodes** [comment]
I'm working on reading PCB's odb++ file and generating the meshed model in python. The spec pdf is huge, it's not an easy work, and I'm improving it. 

Mojo will be more complex than python by design, but it makes things be simple. Keeping simple is hard, and that's why I like mojo. I can see it bringing me to new criteria now. 

Please watch my genuine meme: Maintaining PCBModule.

https://github.com/modularml/mojo/assets/36228443/d3b5e034-11ba-421c-b6ce-66e35383e91d

**Comment by drunkwcodes** [comment]
@gryznar Why did you block me? Though I have no comment on your new issue.

I have introduced myself and gived profound statements, and it's your turn now.
You are trolling.

**Comment by gryznar** [comment]
Sure, sorry about that. What do you think about converting rhis warning to error?

**Comment by lattner** [comment]
For what it's worth, we're not really making the sorts of decisions right now - we're focused on getting the basics in place.  There is a suggestion to completely eliminate 'let' for example, which would define away this topic.  We can track this for evaluation later though, since I think these issues will get sorted out with the lifetime work, which is under active implementation right now.

**Comment by drunkwcodes** [comment]
That suggestion is in this discussion: [let and var Keywords](https://github.com/modularml/mojo/discussions/120).
The breaking news is that Guido van Rossum is also in the discussion!

The `let` and `let mut` in Rust is a new approach to programming, the goodness immunity brings is no need to say.
But the pros and cons needs to be re-consider when the bar of the syntax is high.

I suggest to use `const` instead of `let` in that discussion. And this change will not affect the parallelism of mojo.
Actually, I am amazed by the one-function parallelism in mojo. I need some time to practice and make some tests about it.

Thanks.

**Comment by Marcus-Peterson** [comment]
Please stop rustifying mojo. I get it, rust is great! And I agree with you, I haven't written much rust code but I read some code docs. And I might consider picking up Rust in the future. But this is supposed to be python++ not rust++

**Comment by david-ragazzi** [comment]
> @drunkwcodes you just behave like a usual troll (rejecting ideas without any argument, mocking approach etc.) Keeping your narration serious is just a misunderstanding.

He also is having this unrespectful behavior in #7. I think moderators should put him in his place until he learn respect people.

**Comment by drunkwcodes** [comment]
@david-ragazzi No, I didn't.

**Comment by david-ragazzi** [comment]
I like the concept of keep two keywords to declare mutable/imutable variables. But personally I think it's more intuitive use `const` instead of `let`. So we would have:

- `var`: declare objects which value is variant.
- `const`: declare objects which value is constant.

More intuitive than that, I think it's possible.

**Comment by lattner** [comment]
Closing this out because we aren't tracking things like this in the issue tracker.
